### PR TITLE
Fix logic in target module finding

### PR DESCRIPTION
**Pull request by s-k-yx**
fix the regex in target module finding, refer to #1260 
modify the logic for empty layer pattern, i don't know whether COMMON_LAYERS_PATTERN has special meaning, new logic is: empty ==> all layer pattern is ok

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Thanks for the PR @s-k-yx. I haven't looked at it yet, but could you please run `make style` so that CI can run successfully?

**Comment by s-k-yx** [comment]
> Thanks for the PR @s-k-yx. I haven't looked at it yet, but could you please run `make style` so that CI can run successfully?

sorry for that, the file is reformatted. And there is a dubious test case: https://github.com/huggingface/peft/blob/0f1e9091cc975eb5458cc163bf1843a34fb42b76/tests/test_tuners_utils.py#L65
maybe this should be checked, thanks

File: src/peft/tuners/tuners_utils.py
Comment (BenjaminBossan) [inline src/peft/tuners/tuners_utils.py:507]: Why would you say it's better not to rely on `COMMON_LAYERS_PATTERN` here? If we decide it's really not needed, it should be completely removed from the PEFT code base.

File: src/peft/tuners/tuners_utils.py
Comment (BenjaminBossan) [inline src/peft/tuners/tuners_utils.py:501]: This whole comment is not clear to me, could you please try to word it differently? Note that the comment should make sense as is, even if the reader is not aware of how the code looked like previously.

File: src/peft/tuners/tuners_utils.py
Comment (BenjaminBossan) [inline src/peft/tuners/tuners_utils.py:519]: Probably, instead of using `config.layers_to_transform`, we should use `layer_indexes`.

File: src/peft/tuners/tuners_utils.py
Comment (s-k-yx) [reply src/peft/tuners/tuners_utils.py:504]: I didn't find any other direct usage of COMMON_LAYERS_PATTERN, so I think it won't be harmful. As for the reason why I drop it, I try to avoid changing the output and keep the logic consistent. In the previous code, if the layers_pattern is an empty list, all layer indexes can be matched, as stated in https://github.com/huggingface/peft/blob/482a2a6d9aaa01d534b1240e8c1ab6d346eb278f/tests/test_tuners_utils.py#L59C9-L59C9
In fact, COMMON_LAYERS_PATTERN also doesn't work here. Maybe this is a bug, but I don't change it to keep the output consistent when layers_pattern is an empty list. And now '[]' makes all layers being matched, I think None(default value) should follow this too, even more suitable than '[]'. So I just make them all consistent, None, [], ""(empty string). Then empty layers_pattern means any layer pattern is ok, it's also reasonable. Maybe there should be a warning after the layers_pattern is set.
And torch.nn.ModuleList where use numeric value to be the module name when add modules, i.e. the layer index, so I don't think others will use numeric value as their own module name. Also there can't be '.' in module name, so I think it's safe to just match "[^.]*\.\d+\.".
How to treat empty list in layers_pattern is controversial, here I just follow previous 'bug'.

File: src/peft/tuners/tuners_utils.py
Comment (s-k-yx) [reply src/peft/tuners/tuners_utils.py:498]: Sorry for that, I explain it above, and it will be changed later

File: src/peft/tuners/tuners_utils.py
Comment (s-k-yx) [reply src/peft/tuners/tuners_utils.py:516]: Your are right, layer_indexes should be better. Here I just keep the previous code, I'll change it later

**Comment by s-k-yx** [comment]
Thanks for the helpful suggestions, I'll change test code and comments later. And I'm sorry I didn't make myself clear. I've made some replies, hope it helps.

**Comment by s-k-yx** [comment]
And there is the same problem in layer_indexes, now the new added test case ("foo.bar.7.baz", ["baz"], [], ["bar"], True) will pass(in previous code, it will fail), to be consistent with layers_pattern. Maybe all the empty list should fail, no matter layers_pattern or  layer_indexes

File: src/peft/tuners/tuners_utils.py
Comment (BenjaminBossan) [inline src/peft/tuners/tuners_utils.py:26]: Could you please remove this constant completely from the code base?

File: tests/test_tuners_utils.py
Comment (BenjaminBossan) [inline tests/test_tuners_utils.py:62]: Thanks for adding these examples.

File: tests/test_tuners_utils.py
Comment (BenjaminBossan) [inline tests/test_tuners_utils.py:60]: Also test `""` for `layers_to_transform`?

File: src/peft/tuners/tuners_utils.py
Comment (BenjaminBossan) [inline src/peft/tuners/tuners_utils.py:499]:
```suggestion
            # TODO It's still unclear how empty layers_pattern (None, [], or "") should behave
```

Let's change `fixme` to `TODO`. We use `FIXME` for things that need to be fixed immediately, i.e. before the PR is merged. `TODO` is more long term.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks a lot for improving the PR and your explanations. Those make sense to me. I only have a few smaller change requests left.

> Maybe all the empty list should fail, no matter layers_pattern or layer_indexes

I agree that this could be more sensible, but it could be breaking backwards compatibility of existing code out there, so I'd rather want to avoid it.

File: tests/test_tuners_utils.py
Comment (s-k-yx) [reply tests/test_tuners_utils.py:60]: Due to the type of this arg `layers_to_transform: Optional[Union[List[int], int]])`, these two should be enough.

**Comment by s-k-yx** [comment]
Thanks for the checking, I've modified them accordingly, glad to be of help.

**Review by BenjaminBossan — APPROVED** [review]
Thanks, this LGTM to me now. 

Since this PR has the potential to break existing code, let's wait for another review from @pacman100 or @younesbelkada to ensure that all is good.

File: src/peft/tuners/tuners_utils.py
Comment (younesbelkada) [inline src/peft/tuners/tuners_utils.py:484]:
```suggestion
        target_module_found = re.fullmatch(config.target_modules, key)
```
those are equivalent no?

**Review by younesbelkada — COMMENTED** [review]
Thanks ! I left a single comment, what do you think?

File: src/peft/tuners/tuners_utils.py
Comment (s-k-yx) [reply src/peft/tuners/tuners_utils.py:484]: Thanks for the suggestion, indeed the suggestion is the previous code. I change the code because I see `target_module_found = key in config.target_modules or ...` when the type of `target_modules` is list.  And if the `target_modules` is regex it will be fine, but if the `target_modules` is just the complete layer name and contains special characters in regex, it may cause problem. But for the common layer name like `foo.bar.1.baz`, it doesn't matter, although `.` is special in regex.
  In a word, `target_module_found = re.fullmatch(config.target_modules, key)` should be fine in almost all the cases, but it doesn't hurt to add `(config.target_modules == key) or `.

Comment (younesbelkada) [reply src/peft/tuners/tuners_utils.py:484]: ok looks good to me, thanks for explaining!

**Review by younesbelkada — APPROVED** [review]

**Comment by s-k-yx** [comment]
> Thanks ! I left a single comment, what do you think?

sorry, I found there is the same problem in my change, I shouldn't mix regex and simply compare. It should be better to use `target_module_found = re.fullmatch(config.target_modules, key)` and limit it to regex when `target_modules` is `str`

**Comment by BenjaminBossan** [comment]
> sorry, I found there is the same problem in my change, I shouldn't mix regex and simply compare. It should be better to use `target_module_found = re.fullmatch(config.target_modules, key)` and limit it to regex when `target_modules` is `str`

Just for my understanding, what is the problem exactly? Can you give an example?

**Comment by BenjaminBossan** [comment]
Okay, I think that makes sense. I thought it would be impossible to do this because sub-modules must have a valid Python variable name as identifier, but that's actually not true:

```python
class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.register_module("foo-bar", torch.nn.Linear(1, 2))
        self.register_module("foo/bar$baz^!", torch.nn.Linear(1, 2))

m = MyModule()
getattr(m, "foo-bar")  # works
getattr(m, "foo/bar$baz^!")  # works

config = LoraConfig(target_modules=["foo-bar", "foo/bar$baz^!"])
get_peft_model(m, config)  # works
```

Hopefully, nobody does this in the wild, but we can never know :) Do you think we should have a test for this @s-k-yx?

**Comment by s-k-yx** [comment]
It's always hard to guess user's purpose, telling them how to use makes things simple :) The above test case does nothing wrong, so it should work. Just clarify it in parameter description that it's different when the type of  `target_modules` is `str` and `list`, when it's `str` it should be regex to match the full layer name,  and when it's `list`, the element can be the full layer name or the module name but never regex. Will that be fine? @BenjaminBossan

**Comment by BenjaminBossan** [comment]
> Just clarify it in parameter description that it's different when the type of `target_modules` is `str` and `list`, when it's `str` it should be regex to match the full layer name, and when it's `list`, the element can be the full layer name or the module name but never regex. Will that be fine?

This sound excellent.

**Comment by BenjaminBossan** [comment]
@s-k-yx Would you be so kind as to fix the merge conflict?

**Comment by s-k-yx** [comment]
done @BenjaminBossan

**Pull request merged**
